pyk/src/pyk/kore/client.py

```diff
@@ -0,0 +1,300 @@
+        if reason is StopReason.BRANCHING:
+            raise ValueError(f'Illegal value for reason: {reason}')
+
+        object.__setattr__(self, 'state', state)
+        object.__setattr__(self, 'depth', depth)
+        object.__setattr__(self, 'reason', reason)
+        object.__setattr__(self, 'states', (state,))
+
+    @staticmethod
+    def from_dict(dct: Mapping[str, Any]) -> 'DirectResult':
+        reason = StopReason(dct['reason'])
+
+        state_list = dct['states']
+        if len(state_list) != 1:
+            raise ValueError('Expected states to be a single-element sequence')
+        depth = state_list[0]['depth']
+        state = DirectState(Pattern.from_dict(state_list[0]['state']['term']))
+
+        return DirectResult(state=state, depth=depth, reason=reason)
+
+
+@final
+@dataclass(frozen=True)
+class BranchingResult(ExecuteResult, Iterable[BranchingState]):
+    states: Tuple[BranchingState, ...]
+    depth: int
+
+    reason: InitVar[Literal[StopReason.BRANCHING]]
+
+    def __init__(self, states: Iterable[BranchingState], depth: int):
+        object.__setattr__(self, 'states', tuple(states))
+        object.__setattr__(self, 'depth', depth)
+        object.__setattr__(self, 'reason', StopReason.BRANCHING)
+
+    def __iter__(self) -> Iterator[BranchingState]:
+        return iter(self.states)
+
+    @staticmethod
+    def from_dict(dct: Mapping[str, Any]) -> 'BranchingResult':
+        reason = StopReason(dct['reason'])
+        if reason is not StopReason.BRANCHING:
+            raise ValueError(f'Expected {StopReason.BRANCHING} as reason value, found: {reason}')
+
+        state_list = dct['states']
+        if not state_list:
+            raise ValueError('Expected states to be nonempty')
+
+        assert all(state['depth'] == state_list[0]['depth'] for state in state_list)
+        depth = state_list[0]['depth']
+
+        states = (BranchingState(
+            pattern=Pattern.from_dict(state['term']),
+            substitution=Pattern.from_dict(state['condition']['substitution']),
+            predicate=Pattern.from_dict(state['condition']['predicate']),
+        ) for state in state_list)
+
+        return BranchingResult(states=states, depth=depth)
+
+
+class KoreClient(ContextManager['KoreClient']):
+    _KORE_JSON_VERSION: Final = 1
+
+    _client: JsonRpcClient
+
+    def __init__(self, host: str, port: int):
+        self._client = JsonRpcClient(host, port)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self._client.__exit__(*args)
+
+    def close(self) -> None:
+        self._client.close()
+
+    def _request(self, method: str, **params: Any) -> Dict[str, Any]:
+        try:
+            return self._client.request(method, **params)
+        except JsonRpcError as e:
+            assert e.code != -32602, 'Malformed request'
+            raise KoreClientError(message=e.data['error'], code=e.code, context=e.data['context'])
```

Comment:
I see here the `KoreClientError` but I guess this is only for actually bad input data. Can it be extended to also handle cases like "BackendCrashed" or "SatisfiabilityUnknown"?

Reply:
Whenever the server sends an error response, this exception will be raised. If both server and client implementations are correct, the error code will be either `-32603 Internal error` or `-32000..-32099 Server error`, The client has assertions to check this, as it should never send a malformed request.

Both `BackendCrashed` and `SatisfiabilityUnknown` would fall into the `Server error` category. I'd say however that `SatisfiabilityUnkonwn` should be a `result`, and not an `error` response.

Reply:
Here's an actual stack trace.

```
>>> from pyk.kore.syntax import *
>>> from pyk.kore.client import *
>>> import logging
>>> logging.basicConfig(level=logging.INFO)
>>> client = KoreClient('localhost', 3000)
INFO:pyk.kore.client:Connected to host: localhost:3000
>>> client.execute(Top(SortApp('IntSort')))
INFO:pyk.kore.client:Sending request to localhost:3000: {"jsonrpc": "2.0", "id": 1, "method": "execute", "params": {"state": {"format": "KORE", "version": 1, "term": {"tag": "Top", "sort": {"tag": "SortApp", "name": "IntSort", "args": []}}}}}
INFO:pyk.kore.client:Waiting for response from localhost:3000...
INFO:pyk.kore.client:Received response from localhost:3000: {"error":{"data":{"context":["\\top (<unknown location>)","sort 'IntSort' (<unknown location>)","(<unknown location>)"],"error":"Sort 'IntSort' not defined."},"code":-32002,"message":"Could not verify KORE pattern"},"jsonrpc":"2.0","id":1}

Traceback (most recent call last):
  File "/home/ttoth/git/k/pyk/src/pyk/kore/client.py", line 252, in _request
    return self._client.request(method, **params)
  File "/home/ttoth/git/k/pyk/src/pyk/kore/client.py", line 96, in request
    self._check(data)
  File "/home/ttoth/git/k/pyk/src/pyk/kore/client.py", line 107, in _check
    raise JsonRpcError(**response['error'])
pyk.kore.client.JsonRpcError: Could not verify KORE pattern

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ttoth/git/k/pyk/src/pyk/kore/client.py", line 280, in execute
    result = self._request('execute', **params)
  File "/home/ttoth/git/k/pyk/src/pyk/kore/client.py", line 255, in _request
    raise KoreClientError(message=e.data['error'], code=e.code, context=e.data['context'])
pyk.kore.client.KoreClientError: Sort 'IntSort' not defined.
```

Reply:
I agree with @tothtamas28 on `SatisfiabilityUnknown` being a result. As oth the user errors, currently we have two user error codes, namely:
- -32001 Cancel request unsupported in batch mode
- -32002 Could not verify KORE pattern

I will start adding them to the API doc in a separate section to keep a track of them.

Reply:
Excellent, thank you @goodlyrottenapple . I'll approve this PR, @tothtamas28 feel free to merge when you're ready!